File: python/mirage/commands/builtin/generic/unexpand.py
from collections.abc import AsyncIterator, Awaitable, Callable

from mirage.commands.builtin.utils.stream import _read_stdin_async
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


def _unexpand_line(line: str, tabsize: int, all_spaces: bool) -> str:
    if all_spaces:
        result: list[str] = []
        i = 0
        while i < len(line):
            count = 0
            while i + count < len(line) and line[i + count] == " ":
                count += 1
            if count >= tabsize:
                tabs = count // tabsize
                remainder = count % tabsize
                result.append("\t" * tabs + " " * remainder)
                i += count
            elif count > 0:
                result.append(" " * count)
                i += count
            else:
                result.append(line[i])
                i += 1
        return "".join(result)
    leading = 0
    while leading < len(line) and line[leading] == " ":
        leading += 1
    if leading >= tabsize:
        tabs = leading // tabsize
        remainder = leading % tabsize
        return "\t" * tabs + " " * remainder + line[leading:]
    return line


async def unexpand(
    paths: list[PathSpec],
    *,
    read_bytes: Callable[..., Awaitable[bytes]],
    accessor: object = None,
    stdin: AsyncIterator[bytes] | bytes | None = None,
    tabsize: int = 8,
    all_spaces: bool = False,
) -> tuple[ByteSource | None, IOResult]:
    if paths:
        all_text: list[str] = []
        for p in paths:
            data = (await read_bytes(accessor, p)).decode(errors="replace")
            lines = data.splitlines(True)
            all_text.extend(
                _unexpand_line(ln, tabsize, all_spaces) for ln in lines)
        return "".join(all_text).encode(), IOResult()

    raw = await _read_stdin_async(stdin)
    if raw is None:
        raise ValueError("unexpand: missing operand")
    lines = raw.decode(errors="replace").splitlines(True)
    result = [_unexpand_line(ln, tabsize, all_spaces) for ln in lines]
    return "".join(result).encode(), IOResult()


__all__ = ["unexpand"]

File: typescript/packages/core/src/commands/builtin/generic/unexpand.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { IOResult, materialize, type ByteSource } from '../../../io/types.ts'
import type { PathSpec } from '../../../types.ts'
import type { CommandFnResult, CommandOpts } from '../../config.ts'
import { readStdinAsync } from '../utils/stream.ts'

const ENC = new TextEncoder()
const DEC = new TextDecoder('utf-8', { fatal: false })

function splitLinesKeepEnds(text: string): string[] {
  const lines: string[] = []
  let start = 0
  for (let i = 0; i < text.length; i++) {
    if (text[i] === '\n') {
      lines.push(text.slice(start, i + 1))
      start = i + 1
    }
  }
  if (start < text.length) lines.push(text.slice(start))
  return lines
}

function unexpandLine(line: string, tabsize: number, allSpaces: boolean): string {
  if (allSpaces) {
    const result: string[] = []
    let i = 0
    while (i < line.length) {
      let count = 0
      while (i + count < line.length && line[i + count] === ' ') count += 1
      if (count >= tabsize) {
        const tabs = Math.floor(count / tabsize)
        const remainder = count % tabsize
        result.push('\t'.repeat(tabs) + ' '.repeat(remainder))
        i += count
      } else if (count > 0) {
        result.push(' '.repeat(count))
        i += count
      } else {
        result.push(line[i] ?? '')
        i += 1
      }
    }
    return result.join('')
  }
  let leading = 0
  while (leading < line.length && line[leading] === ' ') leading += 1
  if (leading >= tabsize) {
    const tabs = Math.floor(leading / tabsize)
    const remainder = leading % tabsize
    return '\t'.repeat(tabs) + ' '.repeat(remainder) + line.slice(leading)
  }
  return line
}

export async function unexpandGeneric(
  paths: PathSpec[],
  opts: CommandOpts,
  stream: (p: PathSpec) => AsyncIterable<Uint8Array>,
): Promise<CommandFnResult> {
  const tabsize = typeof opts.flags.t === 'string' ? Number.parseInt(opts.flags.t, 10) : 8
  const allSpaces = opts.flags.a === true
  if (paths.length > 0) {
    const parts: string[] = []
    for (const p of paths) {
      const data = DEC.decode(await materialize(stream(p)))
      for (const ln of splitLinesKeepEnds(data)) parts.push(unexpandLine(ln, tabsize, allSpaces))
    }
    const result: ByteSource = ENC.encode(parts.join(''))
    return [result, new IOResult()]
  }
  const stdinData = await readStdinAsync(opts.stdin)
  if (stdinData === null) {
    return [null, new IOResult({ exitCode: 1, stderr: ENC.encode('unexpand: missing operand\n') })]
  }
  const text = DEC.decode(stdinData)
  const lines = splitLinesKeepEnds(text)
  const result: ByteSource = ENC.encode(
    lines.map((ln) => unexpandLine(ln, tabsize, allSpaces)).join(''),
  )
  return [result, new IOResult()]
}
